proposals/0204-slashable-event-verification.md

```diff
@@ -0,0 +1,297 @@
+                                       // data
+}
+```
+
+We wish to verify that these instructions correspond to
+
+```
+verify(pubkey = node_pubkey, message = shred1.merkle_root, signature = shred1.signature)
+verify(pubkey = node_pubkey, message = shred2.merkle_root, signature = shred2.signature)
+```
+
+We use the deserialized offsets to calculate [\[3\]](#notes) the `pubkey`,
+`message`, and `signature` of each instruction and verify that they correspond
+to the `node_pubkey`, `merkle_root`, and `signature` specified by the shred payload.
+
+If both proof and signer verification succeed, we continue on to store the incident.
+
+---
+
+### Incident reporting
+
+After verifying a successful proof we store the results in a program derived
+address for future use. The PDA is derived using the `node_pubkey`, `slot`, and
+the violation type:
+
+```rust
+let (pda, _) = find_program_address(&[
+  node_pubkey.to_bytes(),
+  slot.to_le_bytes(),
+  ViolationType::DuplicateBlock.to_u8(),
+])
+```
+
+At the moment `DuplicateBlock` is the only violation type but future work will
+add additional slashing types.
+
+If the `pda` account has non-zero lamports, then we abort as the violation has
+already been reported. Otherwise we create the account, with the slashing program
+as the owner. In this account we store the following:
+
+```rust
+struct ProofReport {
+  reporter: Pubkey,              // Fee payer, to allow the account to be closed
+  epoch: Epoch,                  // Epoch in which this report was created
+  pubkey: Pubkey,                // The pubkey of the node that committed the violation
+  slot: Slot,                    // Slot in which the violation occured
+  violation_type: u8,            // The violation type
+  proof: Vec<u8>                 // The serialized proof
+  proof_account: Option<Pubkey>, // Optional account where proof is stored instead
+}
+```
+
+The `DuplicateBlockProofData` is serialized into the `proof` field. This provides
+an on chain trail of the reporting process, since the `proof_account` supplied in
+the `DuplicateBlockProof` account could later be modified.
+
+The `pubkey` is populated with the `node_pubkey`. For future violation types that
+involve votes, this will instead be populated with the vote account's pubkey.
+The work in SIMD-0180 will allow the `node_pubkey` to be translated to a vote account
+if needed.
+
+Note that PDA's can only be created with a 10kb initial size.
+Although not a problem for `DuplicateBlockProofData`, if future proof types require
+more space, we allow the proof to be stored in a separate account, and linked back
+to the PDA using the `proof_account` field.
+
+---
+
+### Closing the incident report
+
+After the slashing violation has been processed by the runtime, the initial fee
+payer may wish to close their `ProofReport` account to reclaim the lamports.
+
+They can accomplish this via the `CloseProofReport` instruction which requires
+2 accounts:
```

Comment:
Sounds like it needs one more account -- the reporter to sign. But I wonder if we can allow this to be totally permissionless by having the reporter provide the lamport destination when they create the report. That way we're sure slash reports are always cleaned up.

Reply:
I like that approach, can definitely see people forgetting to close their reports

Reply:
Could we just send it back to the initial reporter? Or is it worth carrying around a separate address as the destination

Reply:
I'd keep them separate -- it gives extra flexibility and clarity at the cost of an extra 32 bytes, which shouldn't be too bad, especially considering the accounts will get cleaned up

Reply:
separated it here f89a01ba5262bb8b2e79d79f373a84e2dd534d3a